python/ray/dataframe/iterator.py

```diff
@@ -0,0 +1,33 @@
+class PartitionIterator(object):
+    def __init__(self, partitions, func):
+        """PartitionIterator class to define a generator on partitioned data
+
+        Args:
+            partitions ([ObjectID]): Partitions to iterate over
+            func (callable): The function to get inner iterables from
+                each partition
+        """
+        self.partitions = partitions
+        self.curr_partition = -1
+        self.func = func
+        self.iter_cache = iter([])
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.next()
+
+    def next(self):
+        try:
+            n = next(self.iter_cache)
+            return n
+        except StopIteration:
+            self.curr_partition += 1
```

Comment:
The index reference here is needed, as it is used to get the outer `index` or `columns` for that partition.

Reply:
Can you handle the increment outside of this class?

Reply:
The purpose of the `Iterator` is to iterate through each partition, it would be possible to define a function to increment the `partitions` outside of this class, but that would make the code much more complex.

Right now, it checks if there are any items remaining in the current partition, and if not, increments `curr_partition` and gets the next one.